### DISCUSSION TITLE
How do you restrict database access in Kubernetes?

### DISCUSSION THREAD

**Discussion by yorinasub17** [gruntwork-io/knowledge-base#104] [Help] [answered]
I want to know what mechanisms there are to restrict access to a Database from a Kubernetes Pod.

**Comment by gruntwork-discussions** [comment]
You must label this issue with at least one of the following labels before it can be answered:

* [r:terraform-aws-service-catalog](https://github.com/gruntwork-io/terraform-aws-service-catalog)
* [r:terraform-aws-vpc](https://github.com/gruntwork-io/terraform-aws-vpc)
* [r:terraform-aws-architecture-catalog](https://github.com/gruntwork-io/terraform-aws-architecture-catalog)
* [r:terraform-aws-asg](https://github.com/gruntwork-io/terraform-aws-asg)
* [r:terraform-aws-beanstalk](https://github.com/gruntwork-io/terraform-aws-beanstalk)
* [r:terraform-aws-cache](https://github.com/gruntwork-io/terraform-aws-cache)
* [r:terraform-aws-ci](https://github.com/gruntwork-io/terraform-aws-ci)
* [r:terraform-aws-ci-pipeline-example](https://github.com/gruntwork-io/terraform-aws-ci-pipeline-example)
* [r:terraform-aws-cis-service-catalog](https://github.com/gruntwork-io/terraform-aws-cis-service-catalog)
* [r:terraform-aws-data-storage](https://github.com/gruntwork-io/terraform-aws-data-storage)
* [r:terraform-aws-ecs](https://github.com/gruntwork-io/terraform-aws-ecs)
* [r:terraform-aws-eks](https://github.com/gruntwork-io/terraform-aws-eks)
* [r:terraform-aws-elk](https://github.com/gruntwork-io/terraform-aws-elk)
* [r:terraform-aws-kafka](https://github.com/gruntwork-io/terraform-aws-kafka)
* [r:terraform-aws-lambda](https://github.com/gruntwork-io/terraform-aws-lambda)
* [r:terraform-aws-load-balancer](https://github.com/gruntwork-io/terraform-aws-load-balancer)
* [r:terraform-aws-messaging](https://github.com/gruntwork-io/terraform-aws-messaging)
* [r:terraform-aws-mongodb](https://github.com/gruntwork-io/terraform-aws-mongodb)
* [r:terraform-aws-monitoring](https://github.com/gruntwork-io/terraform-aws-monitoring)
* [r:terraform-aws-openvpn](https://github.com/gruntwork-io/terraform-aws-openvpn)
* [r:terraform-aws-sam](https://github.com/gruntwork-io/terraform-aws-sam)
* [r:terraform-aws-security](https://github.com/gruntwork-io/terraform-aws-security)
* [r:terraform-aws-server](https://github.com/gruntwork-io/terraform-aws-server)
* [r:terraform-aws-static-assets](https://github.com/gruntwork-io/terraform-aws-static-assets)
* [r:terraform-aws-zookeeper](https://github.com/gruntwork-io/terraform-aws-zookeeper)
* [r:terragrunt](https://github.com/gruntwork-io/terragrunt)
* [r:terratest](https://github.com/gruntwork-io/terratest)
* [r:repo-copier](https://github.com/gruntwork-io/repo-copier)


[Community Guidelines](https://github.com/gruntwork-io/knowledge-base/discussions/1)
[Getting help](https://gruntwork.io/support/)
[Gruntwork documentation](https://www.gruntwork.io)

**Comment by yorinasub17** [comment, marked as answer]
There are several ways to do this:

- If the database credentials are stored in AWS Secrets Manager, you can restrict which secrets the Pod has access to at the IAM role layer. That is, you setup an IAM policy to only allow the pod to access the secrets it needs and attach it to the IAM role corresponding to the service account.

- If the database credentials are stored in Kubernetes Secrets, then you can manage the RBAC permissions on the service account attached to the Pod to restrict which secrets it has access to mount in.

- We also support the following ways to restrict network access to the data store:
    - If you are using Fargate, you can control which subnets the pods will source it’s IP from on the Fargate Profile. You can take advantage of this to dedicate a special subnet block to Pods in specific namespaces and set up the Database security group to only allow access from that subnet CIDR.

    - If you are using self managed ASGs or managed node groups, you can do something similar to Fargate by setting up ASG node pools for each use case, and restrict the subnets that are used for pods deployed on those nodes.
    
- There are also a few other mechanisms to restrict network access (note that Gruntwork does not yet have modules for configuring these mechanisms):
    - You can assign [custom security groups to your pods](https://docs.aws.amazon.com/eks/latest/userguide/security-groups-for-pods.html) and restrict the network policy for a group of Pods that way.
    - You can setup [Calico](https://docs.aws.amazon.com/eks/latest/userguide/calico.html) and [Network Policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) to setup network firewalls within Kubernetes.
    - You can setup a service mesh solution like [istio](https://istio.io/), and restrict network traffic that way.
